FIVETRAN_CODING_PRINCIPLES.md

```diff
@@ -0,0 +1,192 @@
+- Add docstrings explaining each function’s purpose.
+- Add comments for complex logic and user customization points.
+- Split large functions into smaller helper functions for readability.
+- Use logging judiciously — avoid excessive or redundant logs.
+- Implement error handling that targets specific exceptions.
+- Add retry logic (with exponential backoff) for API requests.
+- Define a complete data model in the `schema()` function, with primary keys and data types.
+- Avoid loading all data into memory — use pagination or streaming.
+- Document pagination, upsert/update/delete, and checkpointing with clear comments.
+- Checkpoint state regularly to resume from the last successful sync and cleanly send data to your warehouse.
+- Refer to the [Connector SDK Best Practices](https://fivetran.com/docs/connectors/connector-sdk/best-practices).
+
+---
+
+## 5. Example development guidelines
+
+### 5.1 Example principles
+
+- Self-explanatory and well-commented code.
+- Must produce working code (can require the SDK Playground).
+- No identifiable customer information should be included.
+
+
+
+### 5.2 Demonstrate one core concept
+
+- Each example should focus on one SDK feature or pattern (e.g., incremental sync, schema discovery, error handling).
+- Split complex connectors into multiple examples.
+
+
+
+### 5.3 Avoid overloading
+
+- Don’t demonstrate all features in one example.
+- Create separate folders/examples, each with a distinct goal.
+- Keep the `update()` method short — use smaller helper methods.
+
+
+### 5.4 Use a clear, self-contained folder
+
+- Use descriptive names like `incremental_sync_example` or `dynamic_schema_example`.
+- Include a simple `README.md` describing:
+  - What the example does.
+  - How to run it.
+
+
+
+### 5.5 Minimal dependencies
+
+- Include only necessary dependencies.
+- Pin dependency versions in `requirements.txt`, `pyproject.toml`, or `poetry.lock`.
+
+
+### 5.6 Code organization
+
+- Keep the code easy to read and navigate.
+- Use good naming instead of excessive commenting.
+- Follow established commenting conventions:
+  - Explain the example’s purpose.
+  - Explain helper functions with inline comments.
+  - Avoid monolithic `update` functions — split logic into helpers.
+
+
+
+### 5.7 Logging
+
+- Log key steps at info level (e.g., `"Fetching page X"`).
+- Avoid logs inside loops or spammy debug logs.
+- Add logs for warnings and severe issues.
+
+
+### 5.8 Error handling
+
+- Catch and raise meaningful, specific exceptions.
+- Avoid generic exception handling.
+- If demonstrating error handling, highlight it in the `README.md`.
+
+
+### 5.9 Testing (optional but encouraged)
+
+- Include a basic test file (e.g., `test_connector.py`).
+- Use `pytest` or similar for validation.
+- Mock external API calls so examples run without real credentials.
+
+### 5.10 Document known limitations
+
+- Note any unhandled edge cases.
+- Provide guidance for contributors on how to extend or enhance the example.
```

Comment:
IMHO, this recommendation is too challenging unless the author has some improvement on their mind but didn't implement it (due to what reasons?). I can't imagine such a case being common.

Reply:
kk I will axe it